File: android/app/src/test/java/com/hisab/app/domain/QuantityTest.kt
package com.hisab.app.domain

import org.junit.Assert.assertEquals
import org.junit.Test

class QuantityTest {
    // Same documented examples as PRD.md section 24 / DECISIONS.md D019,
    // and the same cases used in server/src/domain/quantity.test.ts.

    @Test
    fun `fromDecimal matches the documented examples`() {
        assertEquals(500L, Quantity.fromDecimal(0.5).scaledUnits) // 0.5 kg
        assertEquals(1250L, Quantity.fromDecimal(1.25).scaledUnits) // 1.25 kg
        assertEquals(2500L, Quantity.fromDecimal(2.5).scaledUnits) // 2.5 litre
        assertEquals(3000L, Quantity.fromDecimal(3.0).scaledUnits) // 3 pieces
    }

    @Test
    fun `toDecimal is the inverse of fromDecimal`() {
        assertEquals(1.25, Quantity.fromDecimal(1.25).toDecimal(), 0.0001)
    }

    @Test
    fun `plus and sum work in scaled integer space`() {
        assertEquals(Quantity(1750), Quantity(500) + Quantity(1250))
        assertEquals(Quantity(2000), listOf(Quantity(500), Quantity(500), Quantity(1000)).sum())
    }

    @Test
    fun `a negative quantity delta such as a sale is allowed`() {
        assertEquals(-500L, Quantity(-500).scaledUnits)
    }
}

File: server/src/domain/quantity.test.ts
import { test } from 'node:test'
import assert from 'node:assert/strict'
import {
  quantity,
  quantityFromDecimal,
  quantityToDecimal,
  addQuantity,
  sumQuantity,
} from './quantity.js'

test('quantity rejects non-integer scaled values', () => {
  assert.throws(() => quantity(1.5), TypeError)
})

// The exact examples from PRD.md section 24 and DECISIONS.md D019.
test('quantityFromDecimal matches the documented examples', () => {
  assert.equal(quantityFromDecimal(0.5), 500) // 0.5 kg
  assert.equal(quantityFromDecimal(1.25), 1250) // 1.25 kg
  assert.equal(quantityFromDecimal(2.5), 2500) // 2.5 litre
  assert.equal(quantityFromDecimal(3), 3000) // 3 pieces
})

test('quantityToDecimal is the inverse of quantityFromDecimal', () => {
  assert.equal(quantityToDecimal(quantityFromDecimal(1.25)), 1.25)
})

test('addQuantity and sumQuantity work in scaled integer space', () => {
  assert.equal(addQuantity(quantity(500), quantity(1250)), 1750)
  assert.equal(sumQuantity([quantity(500), quantity(500), quantity(1000)]), 2000)
})

test('a negative quantity_delta (e.g. a sale) is allowed', () => {
  assert.equal(quantity(-500), -500)
})
